### PR TITLE
TRD: trigger and pileup implementation 

### PR DESCRIPTION
**Pull request by jolopezl**
Two things take place here:

- The implementation of signals pileup, and
- a basic trigger for the TRD, with dead time.

We may want to discuss where to keep the trigger. 

To do:

- Take the trigger information to the trigger records object, or
- omit the trigger records and use the digit time information instead, and
- map labels to digits by using and index in the MC label container (for another PR?)

### CODE REVIEW

File: Detectors/TRD/simulation/include/TRDSimulation/Digitizer.h [hunk 1]
```diff
@@ -96,16 +101,19 @@ class Digitizer
...
   // Digitization containers
-  std::vector<HitType> mHitContainer;                       // the container of hits in a given detector
-  std::vector<MCLabel> mMergedLabels;                       // temporary label container
-  std::array<SignalContainer, kNdet> mSignalsMapCollection; // container for caching signals over a timeframe
-  std::array<DigitContainer, kNdet> mDigitsCollection;      // container for caching digits for paralellization
+  std::vector<HitType> mHitContainer;                            // the container of hits in a given detector
+  std::vector<MCLabel> mMergedLabels;                            // temporary label container
+  std::array<SignalContainer, kNdet> mSignalsMapCollection;      // container for caching signals over a timeframe
+  std::array<DigitContainer, kNdet> mDigitsCollection;           // container for caching digits for paralellization
+  std::deque<std::array<SignalContainer, kNdet>> mPileupSignals; // container for piled up signals
```
Comment (sawenzel) [inline Detectors/TRD/simulation/include/TRDSimulation/Digitizer.h:108]: looks like the include for this is missing

File: Detectors/TRD/simulation/src/Digitizer.cxx [hunk 2]
```diff
@@ -86,45 +86,132 @@ void Digitizer::setSimulationParameters()
 
 void Digitizer::clearCollections()
 {
-  for (int det = 0; det < kNdet; ++det) {
-    mSignalsMapCollection[det].clear();
-    mDigitsCollection[det].clear();
-  }
+  std::for_each(mSignalsMapCollection.begin(), mSignalsMapCollection.end(), [](SignalContainer& sc) { sc.clear(); });
+  std::for_each(mDigitsCollection.begin(), mDigitsCollection.end(), [](DigitContainer& dc) { dc.clear(); });
```
Comment (sawenzel) [inline Detectors/TRD/simulation/src/Digitizer.cxx:90]: consider the more readable version:
```
for(auto& dc : mDigitsCollection) {
  dc.clear();
}
```

File: Detectors/TRD/simulation/src/Digitizer.cxx [hunk 3]
```diff
@@ -86,45 +86,132 @@ void Digitizer::setSimulationParameters()
...
   mMergedLabels.clear();
 }
 
+// void Digitizer::flush(DigitContainer& digits, o2::dataformats::MCTruthContainer<MCLabel>& labels)
```
Comment (sawenzel) [inline Detectors/TRD/simulation/src/Digitizer.cxx:94]: any reason for leaving this commented?

**Comment by tdietel** [comment]
Hello Jorge, 

> Two things take place here:
> 
>     * The implementation of signals pileup, and
> 
>     * a basic trigger for the TRD, with dead time.
> 
> 
> We may want to discuss where to keep the trigger.
> 
> To do:
> 
>     * Take the trigger information to the trigger records object, or
> 
>     * omit the trigger records and use the digit time information instead, and

I have not looked at the code in detail, but it seems the trigger processing is very closely tied to the signals container, i.e. you keep a boolean trigger flag within the signals. I would avoid that, and split the logic into three functions:

- Do the hits2signals conversion without any consideration of the trigger and insert all signals into the signals container.

- Create a list of triggers (these are naturally the trigger records). For now this should be done in a very simple way (e.g. constant dead-time after the last trigger, or no dead-time at all). Note that a complete implentation of that would require input from CTP and the trigger detectors, and we would need to run the TRAPsim to determine the data volume that will impact the dead-time. But this level of detail might not be necessary, but let's design the function so that we can easily replace it in the future. 

- Finally, use the trigger records to pick the signals from the signal container and generate the digits trigger by trigger.

As I said, I did not look completely at the code, so some of that might already be there.

>     * map labels to digits by using and index in the MC label container (for another PR?)

Yes, this should probably go to another PR. We could maybe keep the labels in the signal container. This makes the bookkeeping easier, and we will only use the container with MC.

File: Detectors/TRD/simulation/include/TRDSimulation/Digitizer.h (hunk 1 above)
Comment (jolopezl) [reply Detectors/TRD/simulation/include/TRDSimulation/Digitizer.h:108]: I couldn't have caught this on my machine (macOS), for some reason it is not necessary to actually include deque to make it compile. Of course, I will fix this.

**Comment by sawenzel** [comment]
are there any news here?

**Comment by jolopezl** [comment]
> are there any news here?

Now working back on it.

**Comment by jolopezl** [comment]
> are there any news here?

Now, I am working back on it.

File: Detectors/TRD/simulation/src/Digitizer.cxx (hunk 2 above)
Comment (jolopezl) [reply Detectors/TRD/simulation/src/Digitizer.cxx]: This part of the code is removed now.

File: Detectors/TRD/simulation/src/Digitizer.cxx (hunk 3 above)
Comment (jolopezl) [reply Detectors/TRD/simulation/src/Digitizer.cxx]: removed

**Comment by jolopezl** [comment]
Hello @sawenzel, this is still not passing the production test, and I would like to call for help here because I think this is due to a problem with how I am filling the labels. I am adding a comment as a review for this.

File: Detectors/TRD/simulation/src/Digitizer.cxx [hunk 4]
```diff
@@ -84,65 +84,118 @@ void Digitizer::setSimulationParameters()
...
+    // convert signals to ADCs and fill the digit container
+    bool status = convertSignalsToADC(0, fullSignal, digits);
+    if (!status) {
+      LOG(WARN) << "TRD conversion of signals to digits failed";
+    }
+    // get the labels attached to the signals and put it in the label container
+    for (const auto& iter : fullSignal) {
+      labels.addElements(labels.getIndexedSize(), iter.second.labels);
+    }
+  }
```
Comment (jolopezl) [inline Detectors/TRD/simulation/src/Digitizer.cxx:149]: @sawenzel, calling for help here:

If a signal forms after adding all the signals listed for pileup, a digit is produced. Note that `fullSignal` is a container of signals. In a second step, all the MC labels attached to the signal container are dumped to the label container that is sent out to the DPL. At this point, I am not sure if I am using the MCTruthContainer correctly. I am not getting the pileup and embedding analysis passing, but the ADC signal the truly added.

**Comment by jolopezl** [comment]
I said nothing!, tests are passing: 

```Processing /Users/lopez/alice/sw/osx_x86-64/O2/v1.2.0-1/share/macro/analyzeDigitLabels.C("trddigits.root","TRD")...
 TRD 5683 47664 pileup 1 embedding 1 ```

File: Detectors/TRD/simulation/src/Digitizer.cxx (hunk 4 above)
Comment (jolopezl) [reply Detectors/TRD/simulation/src/Digitizer.cxx:149]: Since the test is passing, I guess the implementation is correct.

**Comment by sawenzel** [comment]
Nice progress. The test passes as is in the high-interaction limit.

But there might still be some slight issue: Setting the interaction rate to 1e4 in $O2_ROOT/prodtests/check_embedding_pileup.sh, I now get

TRD 25942 34806 pileup 0 embedding 0

which looks correct for pileup (since events are spaced in time) but not ok for embedding since that shouldn't be affected. Could you cross-check this please?

**Comment by jolopezl** [comment]
> Nice progress. The test passes as is in the high-interaction limit.
> 
> But there might still be some slight issue: Setting the interaction rate to 1e4 in $O2_ROOT/prodtests/check_embedding_pileup.sh, I now get
> 
> TRD 25942 34806 pileup 0 embedding 0
> 
> which looks correct for pileup (since events are spaced in time) but not ok for embedding since that shouldn't be affected. Could you cross-check this please?

Thanks. This is a good catch, indeed. There were missing important cases. For example, events out of the pileup limit were getting lost. The final part of the algorithm relied on looping over the pileup container. This doesn't happen anymore, because we consider both cases now separately, and the behavior looks correct.

**Comment by jolopezl** [comment]
Now, passing the low and high event rate tests locally.

**Comment by sawenzel** [comment]
Thanks. Working correctly now. @tdietel : Will be merged once you approve.

**Pull request merged**